config: prevent SIGHUP from changing non-liveupdatable parameters

andrzej-jackowski-scylladb opened the pull request:
Before this change, it was possible to change non-liveupdatable config
parameter without process restart. This erroneous behavior not only
contradicts the documentation but is potentially dangerous, as various
components theoretically might not be prepared for a change of
configuration parameter value without a restart. The issue came from
a fact that liveupdatability verification check was skipped for default
configuration parameters (those without its initial values
in configuration file during process start).
    
This change:
 - Introduce _initialization_completed member in config_file
 - Set _initialization_completed=true when config file is processed on
   server start
 - Verify config_file's initialization status during config update - if
   config_file was initialized, prevent from further changes of
   non-liveupdatable parameters
 - Implement ScyllaRESTAPIClient::get_config() that obtains a current
    value of given configuration parameter via /v2/config REST API
 - Implement test to confirm that only liveupdatable parameters are
    changed when SIGHUP is sent after configuration file change

Function set_initialization_completed() is called only once in main.cc,
and the effect is expected to be visible in all shards, as a side effect
of cfg->broadcast_to_all_shards() that is called shortly after. The same
technique was already used for enable_3_1_0_compatibility_mode() call.
    
Fixes scylladb/scylladb#5382


No backport - minor fix.

xemul commented on utils/config_file_impl.hh:
```diff
@@ -205,7 +205,10 @@ void utils::config_file::named_value<T>::add_command_line_option(boost::program_
+    if (config_file_initialization_completed && _liveness != liveness::LiveUpdate) {
+        return;
+    }
     if (_source == config_source::SettingsFile && _liveness != liveness::LiveUpdate) {
```
Why is this check not enough to prevent non-live-updateable options from being changed on HUP?

andrzej-jackowski-scylladb replied:
If config parameter is not included in configuration file its source is "default" (utils::config_file::config_source::None). In such case, the check fails, because "_source == config_source::SettingsFile" evaluates to false.

That is also a reason why the SIGHUP change works only once (see step8 in the original https://github.com/scylladb/scylladb/issues/5382 description) - after the first SIGHUP change the source changes from "default" to "config".

The state ("default" vs "config") of the configuration value is visible in system.config table:
```
cassandra@cqlsh> select name, source from system.config;

 name                                                   | source
--------------------------------------------------------+----------
                                 row_cache_keys_to_save |  default
                            strict_is_not_null_in_views |   config
              max_memory_for_unlimited_query_hard_limit |  default
                                         enable_tablets |   config
                          alternator_encryption_options |  default
```

xemul replied:
I see, thanks.

I think that the method should mirror its `set_value(sstring, config_source)` overload and look like `set_value(const YAML::Node&, config_source)`. Then, the `read_from_yaml()` calls it like

```
    cfg.set_value(node.second, re_read_on_hup ? config_source::SettingsFile : config_source::None);
```

to tell if it's the boot-time set-value or re-read-on-sighup. The inner check is to be updated accordingly.

andrzej-jackowski-scylladb replied:
I modified the set_value argument list, to include an argument of config_source type.
I called it "config_source previous_src", because unlike in ":set_value(sstring value, config_source src)", it indicates the previous source of configuration parameter value, not the new one. 

The inner check change is more bold than the previous one (I modified the the old if instead of adding new one). I believe the new inner check is what the previous one was supposed to be from the beginning.

xemul commented on utils/config_file_impl.hh:
```diff
@@ -225,7 +228,10 @@ bool utils::config_file::named_value<T>::set_value(sstring value, config_source
 
 template<typename T>
-future<> utils::config_file::named_value<T>::set_value_on_all_shards(const YAML::Node& node) {
+future<> utils::config_file::named_value<T>::set_value_on_all_shards(const YAML::Node& node, bool config_file_initialization_completed) {
```
It looks like this method is unused and can be removed, instead of patching

andrzej-jackowski-scylladb replied:
I agree, I removed the unused method.

xemul commented on main.cc:
```diff
@@ -867,6 +867,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             // in all supported architectures, and the broadcast_to_all_shards().get() below
             // will apply the required memory barriers anyway.
             ser::gc_clock_using_3_1_0_serialization = cfg->enable_3_1_0_compatibility_mode();
+            cfg->set_initialization_completed();
```
That's not nice to explicitly "lock" config on boot. E.g. -- you forgot to "lock" it in cql_test_env thus leaving a land-mine there. Can db::config::read_from_file should "lock" itself after first invocation?

andrzej-jackowski-scylladb replied:
My intention was to keep the boolean changes that require the memory barriers together.
I moved "_initialization_completed = true" to config_file::read_from_yaml() with a relevant comment.

andrzej-jackowski-scylladb replied:
Moving `_initialization_completed = true` to  to config_file::read_from_yaml broke two unit tests (test_parse_broken and test_commitlog_kms_encryption_with_slow_key_resolve) that called read_from_yaml more than once. I pushed a new version that sets _initialization_completed in [read_from_file](db::config::read_from_file).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14940/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14940/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_config
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3157/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14290/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6277/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14940/)
#### Build Details:
- Duration: 5 hr 17 min
- Builder: spider2.cloudius-systems.com

andrzej-jackowski-scylladb commented:
v2:
 - Remove unused set_value_on_all_shards(const YAML::Node&)
 - Changed set_value(const YAML::Node&, bool) to set_value(const YAML::Node&, config_source) and modified related logic.
 - Remove set_initialization_completed() and set_initialization_completed() call in main.cc, setting _initialization_completed at the end of read_from_yaml instead.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14962/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14962/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_config
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6294/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3174/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14311/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14962/)
#### Failed Tests (2/38153):
- [test_parse_broken](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14962/testReport/junit/boost/config_test/test_parse_broken) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_parse_broken)
- [test_commitlog_kms_encryption_with_slow_key_resolve](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14962/testReport/junit/boost/encryption_at_rest_test/test_commitlog_kms_encryption_with_slow_key_resolve) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_commitlog_kms_encryption_with_slow_key_resolve)
#### Build Details:
- Duration: 6 hr 52 min
- Builder: i-018c6c052294f2404 (m5d.12xlarge)

andrzej-jackowski-scylladb commented:
v5:
 - Moved the newly added `_initialization_completed = true` from `read_from_yaml` to `read_from_file`.
 
Two unit test failed because they used read_from_yaml more than once. The tests are easy to fix, but I think that read_from_yaml is too internal function, and it is better to set _initialization_completed in read_from_yaml to prevent from misuse.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14991/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14991/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_config
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3201/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6321/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14336/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14991/)
#### Build Details:
- Duration: 11 hr
- Builder: i-0be697f763831cee5 (m5ad.8xlarge)
